File: Group-Video/OpenVideoCall-iOS-Objective-C/OpenVideoCall/VideoSession.m
//
//  VideoSession.m
//  OpenVideoCall
//
//  Created by GongYuhua on 2016/9/12.
//  Copyright © 2016年 Agora. All rights reserved.
//

#import "VideoSession.h"
#import "VideoView.h"

@implementation VideoSession
- (void)setIsVideoMuted:(BOOL)isVideoMuted {
    _isVideoMuted = isVideoMuted;
    ((VideoView *)self.hostingView).isVideoMuted = isVideoMuted;
}

+ (instancetype)localSession {
    return [[VideoSession alloc] initWithUid:0];
}

- (instancetype)initWithUid:(NSUInteger)uid {
    if (self = [super init]) {
        self.uid = uid;
        
        self.hostingView = [[VideoView alloc] init];
        self.hostingView.translatesAutoresizingMaskIntoConstraints = NO;
        
        self.canvas = [[AgoraRtcVideoCanvas alloc] init];
        self.canvas.uid = uid;
        self.canvas.view = ((VideoView *)self.hostingView).videoView;
        self.canvas.renderMode = AgoraVideoRenderModeHidden;
    }
    return self;
}


- (void)updateMediaInfo:(CGSize)resolution fps:(NSInteger)fps {
    MediaInfo *info = [[MediaInfo alloc] initWithDimension:resolution fps:fps];
    [((VideoView *)self.hostingView) updateInfo:info];
}

@end

File: Group Video/OpenVideoCall-macOS/OpenVideoCall/VideoSession.swift
//
//  VideoSession.swift
//  OpenVideoCall
//
//  Created by GongYuhua on 3/28/16.
//  Copyright © 2016 Agora. All rights reserved.
//

import Cocoa

class VideoSession: NSObject {
    var uid: UInt = 0
    var hostingView: VideoView!
    var canvas: AgoraRtcVideoCanvas!
    var size: CGSize?
    var mediaInfo = MediaInfo() {
        didSet {
            hostingView?.update(with: mediaInfo)
        }
    }
    var isVideoMuted = false {
        didSet {
            hostingView.isVideoMuted = isVideoMuted
        }
    }
    
    init(uid: UInt) {
        self.uid = uid
        
        hostingView = VideoView(frame: CGRect(x: 0, y: 0, width: 100, height: 100))
        canvas = AgoraRtcVideoCanvas()
        canvas.uid = uid
        canvas.view = hostingView.videoView
        canvas.renderMode = .fit
    }
}

extension VideoSession {
    static func localSession() -> VideoSession {
        return VideoSession(uid: 0)
    }
    
    func updateMediaInfo(resolution: CGSize? = nil, bitRate: Int? = nil, fps: Int? = nil) {
        if let resolution = resolution {
            mediaInfo.resolution = resolution
        }
        
        if let bitRate = bitRate {
            mediaInfo.bitRate = bitRate
        }
        
        if let fps = fps {
            mediaInfo.fps = fps
        }
        
        hostingView.update(with: mediaInfo)
    }
}
